### PR TITLE
Add support for `constants` in Bidirectional wrapper

### PR DESCRIPTION
**Pull request by nisargjhaveri**
Continuing #9248, this PR adds support for passing `constants` to Bidirectional wrapper for RNN.

### CODE REVIEW

**Comment by fchollet** [comment]
Is there a use case for it though?

**Comment by nisargjhaveri** [comment]
@fchollet, I'm trying to implement bidirectional encoder-decoder with attention. I believe passing output of encoder to bidirectional decoder using `constants` will streamline the implementation of attention mechanism.

**Comment by nisargjhaveri** [comment]
@fchollet, I've updated the PR and added some tests. The test fails on CNTK backend, that I don't use. Any ideas what is it about and what is going wrong?

**Comment by fchollet** [comment]
The tests also seem to be failing with TensorFlow: https://travis-ci.org/keras-team/keras/jobs/346241919 

Seems to be an issue with masking?

Regarding CNTK: it seems a `inputs = reverse(inputs, 1)` is triggering a `slice` with an invalid input. What does `inputs` look like at that point?

**Comment by nisargjhaveri** [comment]
I've fixed the failing test with TensorFlow. There was a mistake.

About the failing test with CNTK, I think the problem is with undefined length of timestamps. But isn't that supposed to work or give an proper error message?

Also, what should I do now, test only with defined number of timestamps as the rest of the tests?

**Comment by fchollet** [comment]
@souptc could you please advise about the CNTK errors on this PR? https://travis-ci.org/keras-team/keras/jobs/359322775

**Comment by souptc** [comment]
It is a bug in cntk_backend.py when we need broadcast on the constant. I have a fix at commit:

https://github.com/souptc/keras/commit/01ff7688a27e0a1cecd828f21e0a3294faff2974

Should I send a separate pr for this? or @nisargjhaveri could you cherry-pick this to your pr?

**Comment by nisargjhaveri** [comment]
I've cherry-picked the commit. Tests are passing now! Thanks, @souptc. :D

File: keras/layers/wrappers.py [hunk 1]
```diff
@@ -417,6 +434,48 @@ def call(self, inputs, training=None, mask=None, initial_state=None):
             return [output] + states
         return output
 
+    def _standardize_args(self, inputs, initial_state, constants):
+        """Standardize `__call__` to a single list of tensor inputs.
+
+        This method is taken from `RNN` as it is.
```
Comment (fchollet) [inline keras/layers/wrappers.py]: Could you please split this method to a standalone utility function in `recurrent.py` (named `def _standardize_args`)? The only instance attribute it needs is `_num_constants`, which can be passed as a function argument (`num_constants`).

That way we don't need to duplicate this code, you can just import it from `recurrent.py`.

**Comment by nisargjhaveri** [comment]
Test failed again :/
Also, I might be wrong, but the failed test with CNTK on Python3.6, I don't think it has anything to do with the changes or even CNTK.

**Comment by nisargjhaveri** [comment]
I ran a build again [here](https://travis-ci.org/nisargjhaveri/keras/builds/364488926), and it passes. How can we trigger a rebuild it here?

**Comment by fchollet** [comment]
Closing/reopening triggers a new CI run.

**Comment by nisargjhaveri** [comment]
Now everything apart from CNTK errored, probably due to some network error. :/

**Comment by nisargjhaveri** [comment]
`test_finite_generator_enqueuer_processes` fails with CNTK on python 3.6. It is the same test that failed the first time [here](https://travis-ci.org/keras-team/keras/jobs/364464519), but with different error! No idea what is happening.

**Comment by souptc** [comment]
Error pop from python send/recv call, still network issue?

2018-04-13 2:42 GMT-07:00 Nisarg Jhaveri <notifications@github.com>:

**Comment by nisargjhaveri** [comment]
Rebased on the new master.
